TheSadRogue.Primitives/PropertyChangedEventHelpers.cs

```diff
@@ -0,0 +1,161 @@
+        }
+    }
+
+
+    /// <summary>
+    /// Helper functions useful for implementing events using <see cref="ValueChangedEventArgs{TProperty}"/> as their parameter.
+    /// </summary>
+    public static class PropertyChangedEventHelpers
+    {
+        /// <summary>
+        /// Sets the given field to the new value, and fires the corresponding event just after the value has been changed.
+        /// The value will be properly reverted to the old value if the event handler throws InvalidOperationException.
+        /// </summary>
+        /// <remarks>
+        /// It is fairly common to have an object with a property, such that when the property changes, an event is fired.
+        /// <see cref="IPositionable"/> represents one such interface that is common when dealing with a 2D grid.
+        ///
+        /// Although the implementation of such a property is relatively trivial, there are a few subtle issues that
+        /// can occur:
+        /// - The event might be fired even if the property was set to the same value it is currently
+        /// - If a handler throws something like InvalidOperationException, the value may not be reverted properly.
+        ///
+        /// This function represents a convenient way to implement such a property.  An implementation using this property
+        /// might look like the example below.
+        ///
+        /// Note that this function can work for properties of any type; you could just as easily implement <see cref="IPositionable"/>
+        /// via this interface, for example.
+        ///
+        /// <example>
+        /// <code>
+        /// class MyClass
+        /// {
+        ///     private int _myProperty;
+        ///     public int MyProperty
+        ///     {
+        ///         get => _myProperty;
+        ///         set => this.SafelySetProperty(ref _myProperty, value, MyPropertyChanged);
+        ///     }
+        ///     public event EventHandler&lt;ValueChangedEventArgs&lt;int&gt;&gt;? MyPropertyChanged;
+        /// }
+        /// </code>
+        ///</example>
+        /// </remarks>
+        /// <param name="self" />
+        /// <param name="propertyField">Field to set.</param>
+        /// <param name="newValue">New value to set to given field.</param>
+        /// <param name="changedEvent">Event to fire when change occurs.</param>
+        /// <typeparam name="TObject">Type of the object the property resides on.</typeparam>
+        /// <typeparam name="TProperty">Type of the property.</typeparam>
+        public static void SafelySetProperty<TObject, TProperty>(this TObject self, ref TProperty propertyField, TProperty newValue,
+                                                EventHandler<ValueChangedEventArgs<TProperty>>? changedEvent)
+            where TObject : notnull
+        {
+            // Nothing to do; the value hasn't changed
+            if (propertyField is null)
+            {
+                if (newValue is null) return;
+            }
```

Comment:
Wouldn't this be better with
```suggestion
            if (propertyField is null && newValue is null) return
```

Reply:
It would mess up the else statement.  The entire logic is current this:
```CS
if (propertyField is null)
{
     if (newValue is null) return;
}
else if (propertyField.Equals(newValue)) return;
```

In the case where propertyField is null but newValue is not, the current logic intentionally skips the `Equals` check and any early returns, becuase the values are clearly different.  If I did a compound statement (which would look neater), I'd have to do some additional checking to validate state since the else-if would no longer be associated with only the `propertyField` check.

Reply:
Ahh yes, duh!